### PR TITLE
Closes #2624: Open specific external links in new tab 

### LINKED ISSUE #2624: Add context aware behavior for opening external links in a new tab.
### Is your feature request related to a problem? Please describe.
When looking at a calendar of events, users might be browsing events rather than looking for a specific event. In the current iteration, if someone clicks on a Trellis Event they are taken to a new website kind of unexpectedly and lose the context of the other events on the calendar.

This is a request from Pila as a requirement for the University Calendar.

### Describe the solution you'd like
If an Event is going to take you to a new website (like an imported Trellis Event), then it should open in a new tab, so that people don't lose their place on the calendar.

### Describe alternatives you've considered
An alternative approach would be to stay on the site and display the Event node of the imported event, with an option to click out to events with additional Registration or Session information.

### PR DESCRIPTION
**Pull request by danahertzberg**
## Description

- New branch for the previous issue/2624.
- Includes Chris' changes (not sure if needed now).
- Install new required module with default settings.
- Can have related permissions, but probably better in a separate ticket so specific settings can be exposed to azcontentadmin users rather than all the settings exposed.

## Related issues
Closes #2624 

## How to test

1. Add a PDF to the site
2. Add links to a page
   - PDF on the site
   - Trellis event
   - Google doc
   - Box link
   - OneDrive link
   - Dropbox
3. Test the links to ensure they all open in a new tab/window based on personal setting.

Probo build: https://a5afed90-0208-4d68-9315-65a77c418575.probo.build/

### Example test links

- [Trellis event link](https://events.trellis.arizona.edu/f44lNu67/5a3U3nIwLYK?_gl=1*1ncy4ls*_ga*OTMyMDU0ODAzLjE2OTE0MjM0NjM.*_ga_7PV3540XS3*MTY5NTY3MDQ4NC4zNS4xLjE2OTU2NzA0OTUuNDkuMC4w)
- [Sample Document.pdf](https://17dd72e3-b662-468c-acaf-1d4a01bea572.probo.build/sites/default/files/2023-09/Sample%20Document.pdf)
- [Google doc](https://docs.google.com/document/d/1hYZIm553E7gO_bCxr1K7l0NN_v137k6jNlkUyyzCroY/edit?usp=sharing)
- [Box link](https://arizona.box.com/s/wasg24vebx58n5sv13n9zg68aox59bh0)
- [OneDrive link](https://emailarizona-my.sharepoint.com/:x:/g/personal/danahertzberg_arizona_edu/EcyTb-u8uGpHnmeogdAYRqUBnVZjOPKMxX5v5OB6XuA6Sw?e=KTFRGx)
- [Dropbox](https://www.dropbox.com/scl/fi/8wcc4shset2qkbt3qc6wh/21-UArizona-Dynamic-Template.pptx?rlkey=ms6028rbz9m2nrje7n907o9fg&dl=0)

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_event/az_event_trellis/az_event_trellis.module [hunk 1]
```diff
@@ -79,3 +79,13 @@ function az_event_trellis_migration_plugins_alter(array &$migrations) {
...
+
+/**
+ * Implements hook_link_alter().
+ */
+function az_event_trellis_link_alter(&$variables) {
+  if ($variables['url']->isExternal()) {
+trackleft marked this conversation as resolved.
+    $variables['options']['attributes'] = ['target' => '_blank', 'class' => 'text-decoration-none'];
+  }
+}
```
Comment (danahertzberg) [inline modules/custom/az_event/az_event_trellis/az_event_trellis.module:91]: Is this still required?

File: modules/custom/az_event/az_event_trellis/az_event_trellis.module [hunk 2]
```diff
@@ -79,3 +79,13 @@ function az_event_trellis_migration_plugins_alter(array &$migrations) {
   }
 
 }
+
+/**
+ * Implements hook_link_alter().
+ */
+function az_event_trellis_link_alter(&$variables) {
+  if ($variables['url']->isExternal()) {
+trackleft marked this conversation as resolved.
```
Comment (trackleft) [inline modules/custom/az_event/az_event_trellis/az_event_trellis.module:88]:
```suggestion
```

File: modules/custom/az_event/az_event_trellis/az_event_trellis.module (hunk 1 above)
Comment (trackleft) [inline modules/custom/az_event/az_event_trellis/az_event_trellis.module:91]:
```suggestion

```

File: composer.json [hunk 3]
```diff
@@ -70,6 +70,7 @@
         "drupal/environment_indicator": "4.0.10",
         "drupal/exclude_node_title": "1.4.0",
         "drupal/externalauth": "2.0.3",
+        "drupal/extlink": "^1.7",
```
Comment (joeparsons) [inline composer.json:73]:
```suggestion
        "drupal/extlink": "1.7",
```
We typically pin to a specific release in order to try to keep or builds more deterministic.

File: config/install/extlink.settings.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+extlink_target: true
+extlink_target_no_override: true
+extlink_nofollow: false
+extlink_noreferrer: false
+extlink_follow_no_override: false
+extlink_subdomains: false
+extlink_alert: false
+extlink_alert_text: 'This link will take you to an external web site. We are not responsible for their content.'
+extlink_exclude: ''
+extlink_include: '/sites/default/files/*'
```
Comment (danahertzberg) [inline config/install/extlink.settings.yml:12]:
```suggestion
extlink_include: '/sites/.+/files/.+\.pdf'
```

File: config/install/extlink.settings.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  links: 'fa fa-external-link'
+  mailto: 'fa fa-envelope-o'
+extlink_icon_placement: append
+whitelisted_domains:
+  - 'https://events.trellis.arizona.edu/*'
+  - 'https://arizona.box.com/*'
+  - 'https://docs.google.com/*'
+  - 'https://dropbox.com/*'
+  - 'https://emailarizona-my.sharepoint.com/*'
+
```
Comment (danahertzberg) [inline config/install/extlink.settings.yml:31]:
```suggestion

```

File: config/install/extlink.settings.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+extlink_font_awesome_classes:
+  links: 'fa fa-external-link'
+  mailto: 'fa fa-envelope-o'
+extlink_icon_placement: append
+whitelisted_domains:
+  - 'https://events.trellis.arizona.edu/*'
+  - 'https://arizona.box.com/*'
+  - 'https://docs.google.com/*'
+  - 'https://dropbox.com/*'
+  - 'https://emailarizona-my.sharepoint.com/*'
```
Comment (danahertzberg) [inline config/install/extlink.settings.yml:30]: Editing the file show an extra line, but it doesn't show here.

**Review by sanjanans — APPROVED** [review]
All the links are opening in a new tab as expected. Thanks Dana!

**Comment by danahertzberg** [comment]
Is a db change needed for existing sites?

File: az_quickstart.install [hunk 7]
```diff
@@ -118,3 +118,10 @@ function az_quickstart_update_920701() {
 function az_quickstart_update_920702() {
   \Drupal::service('module_installer')->install(['az_paragraphs_splitscreen']);
 }
+
+/**
+ * Enable extlink module by default.
+ */
+function az_quickstart_update_1020801() {
+  \Drupal::service('module_installer')->install(['extlink']);
+}
```
Comment (danahertzberg) [inline az_quickstart.install:127]: Editing the file show an extra line, but it doesn't show here.

**Review by sanjanans — APPROVED** [review]
The links are behaving as expected. Thanks, Dana.

**Review by sanjanans — APPROVED** [review]

**Comment by trackleft** [comment]
I think we should add a permission 
image

**Review by trackleft — APPROVED** [review]

**Review by camikazegreen — APPROVED** [review]

**Pull request merged**